src/harness/reference_models/propagation/ehata/ehata.py

```diff
@@ -0,0 +1,110 @@
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+"""Simple wrapper module for ITS E-Hata module with WinnForum extensions.
+"""
+
+try:
+  from reference_models.propagation.ehata import ehata_its
+except:
+  raise Exception('E-Hata extension module not available. Please compile it using:\n'
+                '  python setup.py build_ext -i')
+
+
+def SetWinnForumExtensions(on):
+  """Activates/Deactivate the Winnforum extensions.
+  By default they are ON.
+  """
+  ehata_its.SetWinnForumExtensions(on)
+
+def ExtendedHata(its_elev, freq_mhz, height_tx, height_rx, region_code):
+  """Computes the E-Hata propagation path loss.
+
+  Inputs:
+    its_elev : Terrain profile in ITS format from transmitter to receiver:
+               - pfl[0] = number of elevation points - 1
+               - pfl[1] = step size, in meters
+               - pfl[2..N] = elevation above mean sea level, in meters
+    freq_mhz:  frequency (MHz).
+    height_tx: height of transmitter (meters).
+    height_rx: height of receiver (meters).
+    region_code: environment code among:
+            22 = suburban
+            23 or 24 = urban
+            other = rural
+  Returns:
+    the path loss in dB.
+  """
+  return ehata_its.ExtendedHata(its_elev, freq_mhz, height_tx, height_rx, region_code)
+
+
+def MedianBasicPropLoss(freq_mhz, height_tx, height_rx, dist_km, region_code):
+  """Computes the Median Basic propagation loss.
+
+  Differs from ExtendedHata() by not applying the various internal corrections.
+
+  Inputs:
+    freq_mhz:  frequency (MHz).
+    height_tx: height of transmitter (meters).
+    height_rx: height of receiver (meters).
+    dist_km: distance between Tx and Rx (km)
+    region_code: environment code among:
+            22 = suburban
+            23 or 24 = urban
+            other = rural
+  Returns:
+    the pathloss in dB.
+  """
+  return ehata_its.MedianBasicPropLoss(freq_mhz, height_tx, height_rx, dist_km, region_code)
+
+
+def CbsdEffectiveHeights(height_cbsd, its_elev):
+  """Get the CBSD effective height 'h_b'.
+
+  According to Winnforum spec R2-SGN-04. Also same logic
+  as in internal ITS code (after corrections).
+
+  Inputs:
+    height_cbsd: height of the CBSD above terrain (meters).
+    its_elev:  terrain profile in ITS format.
+
+  Returns:
+    the CBSD effective height.
+  """
+
+  np = int(its_elev[0])
+  xi = its_elev[1] * 0.001   # step size of the profile points, in km
+  dist_km = np * xi          # path distance, in km
+  elev_cbsd = its_elev[2]
+
+  if height_cbsd < 20:
+    height_cbsd = 20.
+
+  if dist_km < 3.0:
+    eff_height = height_cbsd
+
+  else: # dist_km >= 3 km
```

Comment:
else: # dist_km >= 3 km
    i_start = 2 + int(3.0 / xi)
    i_end = np + 2
    if dist_km > 15:
      i_end = 2 + int(15.0 / xi)
      dist_km = 15.0
  
  avg_height = sum(its_elev[i_start:i_end+1]) / float(i_end - i_start + 1)
 
This will always pick the elevation profile at a point before the 3.0 km mark and use it in the averaging process.
It should be i_start = 3 + int(3.0 / xi)
 

WINNF-17-I-0095r9 Prop Model Edits said
 
For path lengths greater than 3 km but less than 15 km, hb is the structure height, hcbsd, plus (d-3)/12 times the difference between the terrain elevation at the CBSD antenna and the average terrain elevation over a distance of 3 km to the location of the receiver.

Reply:
Right: initially I actually had the following which meet your comment as well:
```
  i_start = 2 + ceil(3.0 / xi)
```
However the internal C++ eHata implementation computes the effective height with the 'int' method - See PreprocessTerrainPath.cpp:
```
    else if (3.0 <= d__km && d__km <= 15.0)
    {
        i_start = 2 + int(3.0 / xi);
        i_end = np + 2;
         for (int i = i_start; i <= i_end; i++)
            sum = sum + pfl[i];
```
So in order to be consistent with eHata ITS implementation, I decided to use the same logic using the point exactly a 3km or the one just before. I dont think this will make a big numerical difference at the end, but we should definitely choose one or the other:
  - use same method as in ehata
  - follow exactly the winnforum spec (which itself tried to describe the ehata implementation, I guess, but with slightly wrong description).
What do you think ?

Reply:
It does make a small difference. We have seen some corner cases. So as you rightly pointed out to be consistent we should follow the winnforum requirement.
We should use i_start = 2 + ceil(3.0 / xi) and also modify PreprocessTerrainPath.cpp accordingly.

Reply:
Please check the latest changes made: 
modified to use path profile strictly includuded in the 3km-15km range,
both in Python and eHata implementation

Reply:
changes look good.  thanks!